proxy/src/main/java/com/wavefront/agent/listeners/tracing/CustomTracingPortUnificationHandler.java

```diff
@@ -0,0 +1,173 @@
+  }
+
+  @VisibleForTesting
+  public CustomTracingPortUnificationHandler(
+      String handle, TokenAuthenticator tokenAuthenticator, HealthCheckManager healthCheckManager,
+      ReportableEntityDecoder<String, Span> traceDecoder,
+      ReportableEntityDecoder<JsonNode, SpanLogs> spanLogsDecoder,
+      @Nullable Supplier<ReportableEntityPreprocessor> preprocessor,
+      final ReportableEntityHandler<Span, String> handler,
+      final ReportableEntityHandler<SpanLogs, String> spanLogsHandler, Sampler sampler,
+      boolean alwaysSampleErrors, Supplier<Boolean> traceDisabled,
+      Supplier<Boolean> spanLogsDisabled, @Nullable WavefrontSender wfSender,
+      Set<String> traceDerivedCustomTagKeys) {
+    super(handle, tokenAuthenticator, healthCheckManager, traceDecoder, spanLogsDecoder,
+        preprocessor, handler, spanLogsHandler, sampler, alwaysSampleErrors, traceDisabled, spanLogsDisabled);
+    this.wfSender = wfSender;
+    this.discoveredHeartbeatMetrics = new ConcurrentHashMap<>();
+    this.traceDerivedCustomTagKeys = traceDerivedCustomTagKeys;
+
+    if (wfSender != null) {
+      wfInternalReporter = new WavefrontInternalReporter.Builder().
+          prefixedWith("tracing.derived").withSource("custom_tracing").reportMinuteDistribution().
+          build(wfSender);
+      // Start the reporter
+      wfInternalReporter.start(1, TimeUnit.MINUTES);
+    } else {
+      wfInternalReporter = null;
+    }
+  }
+
+  @Override
+  protected void report(Span object, boolean sampleError) {
+    // report converted metrics/histograms from the span
+    String applicationName = NULL_TAG_VAL;
+    String serviceName = NULL_TAG_VAL;
+    String cluster = NULL_TAG_VAL;
+    String shard = NULL_TAG_VAL;
+    String componentTagValue = NULL_TAG_VAL;
+    String isError = "false";
+    if (wfInternalReporter != null) {
+      List<Annotation> annotations = object.getAnnotations();
+      for (Annotation annotation : annotations) {
+        switch (annotation.getKey()) {
+          case APPLICATION_TAG_KEY:
+            applicationName = annotation.getValue();
+            continue;
+          case SERVICE_TAG_KEY:
+            serviceName = annotation.getValue();
+          case CLUSTER_TAG_KEY:
+            cluster = annotation.getValue();
+            continue;
+          case SHARD_TAG_KEY:
+            shard = annotation.getValue();
+            continue;
+          case COMPONENT_TAG_KEY:
+            componentTagValue = annotation.getValue();
+            break;
+          case ERROR_SPAN_TAG_KEY:
+            isError = annotation.getValue();
+            break;
+        }
+      }
+
+      if (applicationName.equals(NULL_TAG_VAL) || serviceName.equals(NULL_TAG_VAL)) {
+        logger.warning("Ingested spans discarded because span application/service name is " +
+            "missing.");
+        discardedSpans.inc();
+        return;
+      }
+
+      if (sampleError || sample(object)) {
+        handler.report(object);
```

Comment:
We'll never get here if wfInternalReporter == null

Reply:
I believe the right approach would be to only disable internal metrics if wfInternalReporter is null, but we still want to validate spans (check for presence of application/service tags) and report them.